dts/arm64/intel_socfpga/intel_socfpga_agilex.dtsi

```diff
@@ -57,6 +57,20 @@
 		label = "clock";
 	};
 
+	/* This periph setting is for setting the MMU region
+	 * to cover the following register map due to
+	 * MMU requirement of 4K alignment
+	 * Uart		<0xffC02000 0x100>
+	 * I2C		<0xffc02800 0x200>
+	 * I2C EMAC	<0xffc02a00 0x300>
+	 * Timer SP	<0xffc03000 0x200>
+	 */
```

Comment:
Hi Carlo, when we tried to set the UART/I2C on themselves, and we enabled assert, there was an error complained that the memory size were not 4k aligned. As the size of UART is only 0x100, and I2C is only 0x200, any suggestion to resolve this?

Reply:
I see, that's a weird granularity indeed (and a pretty weird memory map).

But I'm confused about something about UART for example (looking at `drivers/serial/uart_ns16550.c`): are you setting `UART_NS16550_ACCESS_IOPORT`?

- if you are NOT setting it, the driver is MMU-mapping the IO space for you, so it's not clear why you are MMU mapping again using the syscon.

- if you are setting it, the address of the port is taken directly from the DT and the input / output is being done directly on the physical address retrieved from DT. If you need to MMU map the space, and the driver is working correctly for you, that means that the map is 1:1 between virt and phys.

In the second case this is a problem, because the driver is always assuming that you have 1:1 (because it is reading / writing on the physical address) but this only a special case, and the driver is clearly wrong.

Are you setting `UART_NS16550_ACCESS_IOPORT` ?

Reply:
We are not setting the UART_NS16550_ACCESS_IOPORT. but we still need to add the address map to the mmu_region.c right?

Reply:
No, you don't.

The story goes this way:
- `mmu_region.c` is a legacy way to map a region for all the cases where the driver is not supporting the MMIO device API.
- When the driver is indeed using the MMIO device API (like in the `uart_ns16550` case), the mapping is done by the driver itself and you do not need to add the region to `mmu_region.c`

Reply:
Hi Carlo,

I see, Thanks for clarify Carlo.

In that case, can I say, for the correct implementation, whenever we develop a new driver, we should map the region inside the init function of the new driver instead, not mmu_region.c?

I refer to this link, we can map the driver using DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
https://docs.zephyrproject.org/latest/kernel/drivers/index.html

int my_driver_init(const struct device *dev)
{
      const struct my_driver_config *config = dev->config;

      DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);

Reply:
> In that case, can I say, for the correct implementation, whenever we develop a new driver, we should map the region inside the init function of the new driver instead, not mmu_region.c?

Correct.

> I refer to this link, we can map the driver using DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);

Yes, exactly.

Reply:
Hi Carlo,

I look at the example of defining multiple mmio-region at the driver also dts this dosent seems right for me to implement at the qspi driver.

For our QSPI driver, it is defined like this in the dts, 
        qspi: qspi@ff8d2000 {
                compatible = "cdns,qspi-nor";
                reg = <0xff8d2000 0x100>,
                      <0xff900000 0x100000>;
                interrupt-parent = <&gic>;
                interrupts = <GIC_SPI 4 IRQ_TYPE_LEVEL
                              IRQ_DEFAULT_PRIORITY>;
                clock-frequency = <50000000>;
                label = "qspi";
                status = "disabled";
        };


we got two region, 0xff8d200 for the configuration register of qspi and 0xff900000 for the data register of qspi.
i was looking for the example at tests/kernel/device/src/mmio.c, it is defining two nodes for the region (corge and grault) but this would need to create for 2 driver instance. 

what i want is using 1 driver instance to define 2 region, qspi config register and qspi data register.

the only way i can think of is at the qspi driver side, I will be using the using the MMIO-Macro to map the qspi config register and using the device_map function to map the qspi data register. Let me know if this implementation is okay.

Reply:
Uhm, let me prepare a patch for that.

Reply:
Opened #46433